File: study/ts/script/scene/view/singleRace/SingleRaceGuessServerCell.ts
const { ccclass, property } = cc._decorator;

import CommonButtonLevel1Normal from '../../../ui/component/CommonButtonLevel1Normal'

import CommonHeroIcon from '../../../ui/component/CommonHeroIcon'
import { handler } from '../../../utils/handler';
import CommonListItem from '../../../ui/component/CommonListItem';
import { TextHelper } from '../../../utils/TextHelper';
import { UserDataHelper } from '../../../utils/data/UserDataHelper';
import { Path } from '../../../utils/Path';
import { G_UserData } from '../../../init';
import { SingleRaceConst } from '../../../const/SingleRaceConst';
import UIHelper from '../../../utils/UIHelper';
import { UIPopupHelper } from '../../../utils/UIPopupHelper';
import { UserDetailData } from '../../../data/UserDetailData';

@ccclass
export default class SingleRaceGuessServerCell extends CommonListItem {

    @property({
        type: cc.Node,
        visible: true
    })
    _resourceNode: cc.Node = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textServer: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textPower: cc.Label = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageMark: cc.Sprite = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textCount: cc.Label = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _node1: cc.Node = null;

    @property({
        type: CommonHeroIcon,
        visible: true
    })
    _nodeIcon1: CommonHeroIcon = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageOfficial1: cc.Sprite = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textName1: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textPlayerPower1: cc.Label = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _node2: cc.Node = null;

    @property({
        type: CommonHeroIcon,
        visible: true
    })
    _nodeIcon2: CommonHeroIcon = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageOfficial2: cc.Sprite = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textName2: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textPlayerPower2: cc.Label = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _node3: cc.Node = null;

    @property({
        type: CommonHeroIcon,
        visible: true
    })
    _nodeIcon3: CommonHeroIcon = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageOfficial3: cc.Sprite = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textName3: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textPlayerPower3: cc.Label = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _node4: cc.Node = null;

    @property({
        type: CommonHeroIcon,
        visible: true
    })
    _nodeIcon4: CommonHeroIcon = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageOfficial4: cc.Sprite = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textName4: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textPlayerPower4: cc.Label = null;

    @property({
        type: CommonButtonLevel1Normal,
        visible: true
    })
    _buttonVote: CommonButtonLevel1Normal = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageVoted: cc.Sprite = null;
    _data: any;

    onCreate() {
        this._buttonVote.addClickEventListenerEx(handler(this, this._onButtonVoteClicked));

        this._data = null;
        for (var i = 1; i <= 4; i++) {
            this['_nodeIcon' + i].onLoad();
            this['_nodeIcon' + i].setTouchEnabled(true);
            this['_nodeIcon' + i].setCallBack(handler(this, this['_onClickIcon' + i]));
        }
    }
    updateUI(index, datas) {
        var data = datas[0];
        this._data = data;
        var mySupportId = datas[1];
        var supportNum = datas[2];
        var markRes = datas[3];
        var textButton = datas[4];
        this._textServer.string = (data.getServer_name());
        this._textPower.string = (TextHelper.getAmountText1(data.getPower()));
        this.updateCount(supportNum);
        this.updateVote(mySupportId);
        var users = data.getUserDatas();
        for (var i = 1; i <= 4; i++) {
            var user = users[i-1];
            if (user) {
                this['_node' + i].active = (true);
                var [coverId, , limitLevel, limitRedLevel] = user.getCovertIdAndLimitLevel();
                var [officialName, officialColor, officialInfo] = UserDataHelper.getOfficialInfo(user.getOfficer_level());
                this['_nodeIcon' + i].updateUI(coverId, null, limitLevel, limitRedLevel);
                UIHelper.loadTexture(this['_imageOfficial' + i], Path.getTextHero(officialInfo.picture));
                this['_textName' + i].string = (user.getUser_name());
                this['_textName' + i].node.color = (officialColor);
                UIHelper.updateTextOfficialOutline(this['_textName' + i].node, user.getOfficer_level());
                this['_textPlayerPower' + i].string = (TextHelper.getAmountText1(user.getPower()));
            } else {
                this['_node' + i].active = (false);
            }
        }
        UIHelper.loadTexture(this._imageMark, markRes);
        this._buttonVote.setString(textButton);
    }
    updateVote(mySupportId) {
        var status = G_UserData.getSingleRace().getStatus();
        var isInStatus = status == SingleRaceConst.RACE_STATE_PRE;
        var isVoted = mySupportId > 0;
        var thisVoted = this._data.getServer_id() == mySupportId;
        this._buttonVote.node.active = (!isVoted && isInStatus);
        this._imageVoted.node.active = (thisVoted);
    }
    updateCount(supportNum) {
        this._textCount.string = (supportNum);
    }
    _onButtonVoteClicked() {
        this.dispatchCustomCallback(1);
    }
    getDataId() {
        return this._data.getServer_id();
    }
    _onClickIcon1() {
        if (this._data) {
            var users = this._data.getUserDatas();
            var user = users[0];
            if (user) {
                this._popupDetailInfo(user);
            }
        }
    }
    _onClickIcon2() {
        if (this._data) {
            var users = this._data.getUserDatas();
            var user = users[1];
            if (user) {
                this._popupDetailInfo(user);
            }
        }
    }
    _onClickIcon3() {
        if (this._data) {
            var users = this._data.getUserDatas();
            var user = users[2];
            if (user) {
                this._popupDetailInfo(user);
            }
        }
    }
    _onClickIcon4() {
        if (this._data) {
            var users = this._data.getUserDatas();
            var user = users[3];
            if (user) {
                this._popupDetailInfo(user);
            }
        }
    }
    _popupDetailInfo(user) {
        var userId = user.getUser_id();
        var userDetailData = G_UserData.getSingleRace().getUserDetailInfoWithId(userId);
        var power = user.getPower();
        if (userDetailData) {
            UIPopupHelper.popupUserDetailInfor(userDetailData, power);
        } else {
            var pos = G_UserData.getSingleRace().getPosWithUserIdForGuess(userId);
            G_UserData.getSingleRace().c2sGetSingleRacePositionInfo(pos);
        }
    }

}

File: study/lua/src/app/scene/view/singleRace/SingleRaceGuessServerCell.lua

local ListViewCellBase = require("app.ui.ListViewCellBase")
local SingleRaceGuessServerCell = class("SingleRaceGuessServerCell", ListViewCellBase)
local UserDataHelper = require("app.utils.UserDataHelper")
local TextHelper = require("app.utils.TextHelper")
local SingleRaceConst = require("app.const.SingleRaceConst")

function SingleRaceGuessServerCell:ctor()
	local resource = {
		file = Path.getCSB("SingleRaceGuessServerCell", "singleRace"),
		binding = {
			_buttonVote = {
				events = {{event = "touch", method = "_onButtonVoteClicked"}}
			},
		}
	}
	SingleRaceGuessServerCell.super.ctor(self, resource)
end

function SingleRaceGuessServerCell:onCreate()
	self._data = nil
	local size = self._resourceNode:getContentSize()
	self:setContentSize(size.width, size.height)
	for i = 1, 4 do
		self["_nodeIcon"..i]:setTouchEnabled(true)
		self["_nodeIcon"..i]:setCallBack(handler(self, self["_onClickIcon"..i]))
		self["_imageOfficial"..i]:ignoreContentAdaptWithSize(true)
	end
end

function SingleRaceGuessServerCell:update(data, index, mySupportId, supportNum, markRes, textButton)
	self._data = data
	self._textServer:setString(data:getServer_name())
	self._textPower:setString(TextHelper.getAmountText1(data:getPower()))
	self:updateCount(supportNum)
	self:updateVote(mySupportId)
	local users = data:getUserDatas()
	for i = 1, 4 do
		local user = users[i]
		if user then
			self["_node"..i]:setVisible(true)
			local coverId, limitLevel, limitRedLevel = user:getCovertIdAndLimitLevel()
			local officialName, officialColor,officialInfo = UserDataHelper.getOfficialInfo(user:getOfficer_level())
			self["_nodeIcon"..i]:updateUI(coverId, nil, limitLevel, limitRedLevel)
			self["_imageOfficial"..i]:loadTexture(Path.getTextHero(officialInfo.picture))
			self["_textName"..i]:setString(user:getUser_name())
			self["_textName"..i]:setColor(officialColor)
			require("yoka.utils.UIHelper").updateTextOfficialOutline(self["_textName"..i], user:getOfficer_level())
			self["_textPlayerPower"..i]:setString(TextHelper.getAmountText1(user:getPower()))
		else
			self["_node"..i]:setVisible(false)
		end
	end
	self._imageMark:loadTexture(markRes)
	self._buttonVote:setString(textButton)
end

function SingleRaceGuessServerCell:updateVote(mySupportId)
	local status = G_UserData:getSingleRace():getStatus()
	local isInStatus = status == SingleRaceConst.RACE_STATE_PRE
	local isVoted = mySupportId > 0 --投过了
	local thisVoted = self._data:getServer_id() == mySupportId --此项已投
	self._buttonVote:setVisible(not isVoted and isInStatus)
	self._imageVoted:setVisible(thisVoted)
end

function SingleRaceGuessServerCell:updateCount(supportNum)
	self._textCount:setString(supportNum)
end

function SingleRaceGuessServerCell:_onButtonVoteClicked()
	self:dispatchCustomCallback(1)
end

function SingleRaceGuessServerCell:getDataId()
	return self._data:getServer_id()
end

function SingleRaceGuessServerCell:_onClickIcon1()
	if self._data then
		local users = self._data:getUserDatas()
		local user = users[1]
		if user then
			self:_popupDetailInfo(user)
		end
	end
end

function SingleRaceGuessServerCell:_onClickIcon2()
	if self._data then
		local users = self._data:getUserDatas()
		local user = users[2]
		if user then
			self:_popupDetailInfo(user)
		end
	end
end

function SingleRaceGuessServerCell:_onClickIcon3()
	if self._data then
		local users = self._data:getUserDatas()
		local user = users[3]
		if user then
			self:_popupDetailInfo(user)
		end
	end
end

function SingleRaceGuessServerCell:_onClickIcon4()
	if self._data then
		local users = self._data:getUserDatas()
		local user = users[4]
		if user then
			self:_popupDetailInfo(user)
		end
	end
end

function SingleRaceGuessServerCell:_popupDetailInfo(user)
	local userId = user:getUser_id()
	local userDetailData = G_UserData:getSingleRace():getUserDetailInfoWithId(userId)
	local power = user:getPower()
	if userDetailData then
		local popup = require("app.ui.PopupUserDetailInfo").new(userDetailData, power)
	    popup:setName("PopupUserDetailInfo")
	    popup:openWithAction()
	else
		local pos = G_UserData:getSingleRace():getPosWithUserIdForGuess(userId)
		G_UserData:getSingleRace():c2sGetSingleRacePositionInfo(pos)
	end
end

return SingleRaceGuessServerCell